### PR TITLE
Add repository url check

### PR DESCRIPTION
**Pull request by SLdragon**
Some cases:
```
aa/aaaAdd:888
invalid reference format: repository name must be lowercase

aaa/ddd/ddcc/eeAaa/aaadd:888
invalid reference format: repository name must be lowercase

AAA.dd*d/aa/bb:latest
invalid reference format

localHost/aa/bb:latest
invalid reference format: repository name must be lowercase

aaa:*ddd
invalid reference format

aaa/dddd/ddd/aa.aa 
valid

aa:88/aa/bb
valid

AAA.ddd/aa/bb:latest
valid

localhost/aa/bb:latest 
valid
```

### CODE REVIEW

File: src/common/utility.ts [hunk 1]
```diff
@@ -566,6 +566,47 @@ export class Utility {
...
+        const tag: string = match[3];
+
+        // if host name do not contain ":.", then it would be treated as repository name
+        if (hostName && !/[:.]/.test(hostName)) {
+            repositoryName = hostName  + "/" + repositoryName;
+            hostName = null;
+        }
+
+        if (hostName && /[^a-zA-Z0-9\.\-:]/.test(hostName)) {
+            return "Repository host name contains invalid characters";
```
Comment (adashen) [inline src/common/utility.ts]: instead of show invalid. how about show what is valid characters?

File: src/common/utility.ts [hunk 2]
```diff
@@ -566,6 +566,47 @@ export class Utility {
...
+        }
+
+        if (/[^a-z0-9\._\-\/]+/.test(repositoryName)) {
+            return "Repository name contains invalid characters";
```
Comment (adashen) [inline src/common/utility.ts]: same as above

File: src/common/utility.ts [hunk 3]
```diff
@@ -566,6 +566,47 @@ export class Utility {
...
+        }
+
+        if (tag) {
+            if (tag.length > 128) {
```
Comment (adashen) [inline src/common/utility.ts]: the length is just for tag or the whole? (reposity+tag)

File: src/common/utility.ts [hunk 4]
```diff
@@ -566,6 +566,47 @@ export class Utility {
...
+                return "The maximum length of tag is 128";
+            }
+
+            if (/[^a-zA-Z0-9\._\-\/]+/.test(tag)) {
+                return "Tag contains invalid characters";
```
Comment (adashen) [inline src/common/utility.ts]: same as above

File: src/common/utility.ts (hunk 1 above)
Comment (SLdragon) [reply src/common/utility.ts]: OK, thanks, it's more reasonable

File: src/common/utility.ts (hunk 3 above)
Comment (SLdragon) [reply src/common/utility.ts]: just for tag

File: src/common/utility.ts [hunk 5]
```diff
@@ -566,6 +566,47 @@ export class Utility {
         return path.join(solutionPath, Constants.vscodeFolder, Constants.vscodeSettingsFile);
     }
 
+    public static async validateRepositoryUrl(repositoryUrl: string): Promise<string | undefined> {
+        if (!repositoryUrl) {
+            return "Repository url could not be empty";
+        }
+
+        const match: RegExpMatchArray = repositoryUrl.match(/^(?:([^\/]+)\/)?([^:]+)(?::(.+))?$/);
```
Comment (adashen) [inline src/common/utility.ts]: where did we get the regex or write by yourself?

Comment (adashen) [reply src/common/utility.ts]: Can we just get the first part before "/" like what we did in getting address?

Comment (SLdragon) [reply src/common/utility.ts]: myself, this regex is to divide the repository url to three part: hostName, repository name and tag.
However, the validate function cannot cover all invalid cases such as aa/bb///ccdd

File: src/common/utility.ts (hunk 2 above)
Comment (SLdragon) [reply src/common/utility.ts]: updated

File: src/common/utility.ts (hunk 4 above)
Comment (SLdragon) [reply src/common/utility.ts]: updated

File: src/common/utility.ts (hunk 5 above)
Comment (SLdragon) [reply src/common/utility.ts]: > Can we just get the first part before "/" like what we did in getting address?

Yes, the first part of this regex : (?:([^\/]+)\/)?  is to get first part before "/" just like we use indexOf("/")

**Pull request merged**